---
title: "Gaudin_QG_Project_Rcode"
author: "Paul Gaudin"
date: "4/30/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


######Import phenotypes and genotypes into tables
```{r}
phenotypes_table <- read.csv("data_files/phenotypes.csv", 
    header = TRUE, row.names = 1)

genotypes_table <- read.csv("data_files/genotypes.csv", 
    header = TRUE, row.names = 1)
````

Check if rownames match 
```{r}
#rownames(phenotypes_table) == rownames(genotypes_table)
```

Get number of SNPs
```{r}
N_SNPs <- ncol(genotypes_table)
```
Get sample size 
```{r}
sample_size <- nrow(genotypes_table)
```

```{r}
Xa_matrix <- genotypes_table - 1;
Xd_matrix <- 1 - 2*abs(Xa_matrix)
```

```{r}
#write.csv(Xa_matrix, file = "Xa_matrix.csv")
#write.csv(Xd_matrix, file = "Xd_matrix.csv")

```



```{r}
pval_calculator <- function(pheno_input, xa_input, xd_input){
    n_samples <- length(xa_input)
    
    X_mx <- cbind(1,xa_input,xd_input)
    
    MLE_beta <- ginv(t(X_mx) %*% X_mx) %*% t(X_mx) %*% pheno_input
    y_hat <- X_mx %*% MLE_beta
  
    SSM <- sum((y_hat - mean(pheno_input))^2)
    SSE <- sum((pheno_input - y_hat)^2)
  
    df_M <- 2
    df_E <- n_samples - 3 
  
    MSM <- SSM / df_M
    MSE <- SSE / df_E
    
    Fstatistic <- MSM / MSE
  
    pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)
    
    return(pval)
}
```

```{r}
library(MASS)
p_value_table <- matrix(, nrow = ncol(phenotypes_table), ncol = N_SNPs)

for(pheno_col in 1:ncol(phenotypes_table)){
  pvals <- rep(0,ncol(Xa_matrix))
  for(i in 1:ncol(Xa_matrix)){
    pvals[i] <- pval_calculator(phenotypes_table[,pheno_col], Xa_matrix[,i], Xd_matrix[,i])
  }
  p_value_table[pheno_col,] <- pvals
}

```

```{r}
#write.csv(p_value_table, file = "p_value_table.csv")
```

```{r}

library(ggplot2)

library(ggfortify)

pca_matrix <- read.csv("data_files/genotypes.csv", 
    header = TRUE, row.names = 1)

covars <- read.csv("data_files/covars.csv", 
    header = TRUE, row.names = 1)

```

```{r}
pca_matrix[,ncol(genotypes_table) + 1] <- covars$Population
colnames(pca_matrix)[ncol(genotypes_table) + 1] <- "Population"
autoplot(prcomp(genotypes_table), data = pca_matrix, colour = 'Population')
```

```{r}
pca_matrix[,ncol(genotypes_table) + 2] <- covars$Sex
colnames(pca_matrix)[ncol(genotypes_table) + 2] <- "Sex"
autoplot(prcomp(genotypes_table), data = pca_matrix, colour = "Sex")
```


```{r}
#autoplot(prcomp(Xa_matrix), data = pca_matrix, colour = "Population")

```

TSI = -1
CEU = 0
GBR = 1
FIN = 2

```{r}

covar_pheno_vals <- rep(0,length(covars$Population))

for(i in 1:length(covars$Population)){
  if(covars$Population[i] =="TSI"){
    covar_pheno_vals[i] <- -1
  }
  else if(covars$Population[i] =="CEU"){
    covar_pheno_vals[i] <- 0
  }
  else if(covars$Population[i] =="GBR"){
    covar_pheno_vals[i] <- 1
  }
  else{
    covar_pheno_vals[i] <- 2
  }
}

```


```{r}
pval_calculator_covar <- function(xa_input, pheno_input, z_input){ 
  xa_input <- xa_input - 1
  xd_input <- 1 - 2*abs(xa_input)
  n_samples <- length(xa_input)
  Z_mx <- cbind(1,z_input)
  XZ_mx <- cbind(1,xa_input,xd_input,z_input)
  MLE_beta_theta0 <- ginv(t(Z_mx) %*% Z_mx) %*% t(Z_mx) %*% pheno_input 
  MLE_beta_theta1 <- ginv(t(XZ_mx) %*% XZ_mx) %*% t(XZ_mx) %*% pheno_input 
  y_hat_theta0 <- Z_mx %*% MLE_beta_theta0
  y_hat_theta1 <- XZ_mx %*% MLE_beta_theta1
  SSE_theta0 <- sum((pheno_input - y_hat_theta0)^2)
  SSE_theta1 <- sum((pheno_input - y_hat_theta1)^2)
  df_M <- 2
  df_E <- n_samples - 3
  Fstatistic <- ((SSE_theta0-SSE_theta1)/df_M) / (SSE_theta1/df_E)
  pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)
  return(pval)
}
```

```{r}
library(MASS)
p_value_cov_table <- matrix(, nrow = ncol(phenotypes_table), ncol = N_SNPs)

for(pheno_col in 1:ncol(phenotypes_table)){
  #print(pheno_col)
  #pvals <- rep(0,ncol(Xa_matrix))
  for(i in 1:ncol(Xa_matrix)){
    p_value_cov_table[pheno_col,i] <- pval_calculator_covar(Xa_matrix[,i], phenotypes_table[,pheno_col], covar_pheno_vals)
  }
  #p_value_cov_table[pheno_col,] <- pvals
}

```

```{r}
create_qq <- function(pvals){
  N = length(pvals)
  log_sorted_pvals = sort(-log10(pvals))
  N_vector = sort(-log10(seq(0, 1, by = 1/N)))[1:N]
  pval_qqDf <- data.frame(pvals = log_sorted_pvals, N_vector = N_vector)
  qq_plot <- ggplot(pval_qqDf, aes(N_vector, log_sorted_pvals)) + geom_point() + 
    geom_abline(intercept = 0, slope = 1, color = "red")
  return(qq_plot)
}
```

```{r}

qq_1 <- create_qq(p_value_table[1,]) + labs(y = "Computed PVals", x = "")
qq_2 <- create_qq(p_value_table[2,]) + labs(y = "", x = "")
qq_3 <- create_qq(p_value_table[3,]) + labs(y = "", x = "Theor. Pvals")
qq_4 <- create_qq(p_value_table[4,]) + labs(y = "", x = "")
qq_5 <- create_qq(p_value_table[5,]) + labs(y = "", x = "")


```

```{r fig.height = 1, fig.width=5}
library(ggpubr)
theme_set(theme_pubr())

qq_plots_no_cov <- ggarrange(qq_1, qq_2, qq_3, qq_4, qq_5,
                    labels = c("1", "2", "3", "4", "5"),
                    ncol = 5, nrow = 1)

qq_plots_no_cov

```


```{r}
qq_1_cov <- create_qq(p_value_cov_table[1,]) + labs(y = "Computed PVals", x = "")
qq_2_cov <- create_qq(p_value_cov_table[2,]) + labs(y = "", x = "")
qq_3_cov <- create_qq(p_value_cov_table[3,]) + labs(y = "", x = "Theor. Pvals")
qq_4_cov <- create_qq(p_value_cov_table[4,]) + labs(y = "", x = "")
qq_5_cov <- create_qq(p_value_cov_table[5,]) + labs(y = "", x = "")
```

```{r fig.height = 1, fig.width=5}

qq_plots_with_cov <- ggarrange(qq_1_cov, qq_2_cov, qq_3_cov, qq_4_cov, qq_5_cov,
                    labels = c("1", "2", "3", "4", "5"),
                    ncol = 5, nrow = 1)

qq_plots_with_cov
```

```{r}
BF_correction_value <- 0.05/N_SNPs #1e-06

column_names <- colnames(genotypes_table)

pheno_1_most_sig <- c()

for(col in 1:ncol(p_value_table)){
  if(p_value_table[1,col] < BF_correction_value){
    pheno_1_most_sig <- c(pheno_1_most_sig, column_names[col])
  }
}

pheno_2_most_sig <- c()

for(col in 1:ncol(p_value_table)){
  if(p_value_table[2,col] < BF_correction_value){
    pheno_2_most_sig <- c(pheno_2_most_sig, column_names[col])
  }
}

pheno_2_most_sig

pheno_3_most_sig <- c()

for(col in 1:ncol(p_value_table)){
  if(p_value_table[3,col] < BF_correction_value){
    pheno_3_most_sig <- c(pheno_3_most_sig, column_names[col])
  }
}

pheno_3_most_sig

```


```{r}

library(biomaRt)
mart.snp <- useMart(biomart = "ENSEMBL_MART_SNP", dataset="hsapiens_snp", host = "www.ensembl.org", ensemblRedirect = FALSE)

getENSG <- function(rs, mart = mart.snp) {
  results <- getBM(attributes = c("refsnp_id", "ensembl_gene_stable_id"),
                   filters    = "snp_filter", values = rs, mart = mart)
  #print(results)
  return(results)
}

pheno_2_ids <- getENSG(pheno_2_most_sig, mart.snp)[,2];

mart = useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", host = "www.ensembl.org")

dat_for_pheno_2 = getBM(
  values = c(pheno_2_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

pheno_2_ids
dat_for_pheno_2
```


```{r}
pheno_3_ids <- getENSG(pheno_3_most_sig, mart.snp)[,2];

dat_for_pheno_3 = getBM(
  values = c(pheno_3_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

pheno_3_ids
dat_for_pheno_3

```



```{r}
pheno_1_ids <- getENSG(pheno_1_most_sig, mart.snp)[,2];

dat_for_pheno_1 = getBM(
  values = c(pheno_1_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

pheno_1_ids
dat_for_pheno_1

```


```{r}
SNP_info <- read.csv("data_files/SNP_info.csv", 
    header = TRUE)
```

```{r}

p_table_with_chrom_number$chromosome <- SNP_info$chromosome

p_table_with_chrom_number$pheno_1 <- p_value_table[1,]

p_table_with_chrom_number$pheno_2 <- p_value_table[2,]

p_table_with_chrom_number$pheno_3 <- p_value_table[3,]
```

```{r}
#library(ggplot2)
SNPs <- c(1:N_SNPs)
neg_log_ps <- -log10(p_table_with_chrom_number$pheno_1)
chromosomes <- factor(p_table_with_chrom_number$chromosome)
df <- data.frame(SNPs,neg_log_ps,p_table_with_chrom_number$chromosome)
ggplot(df, aes(x=SNPs, y=neg_log_ps, group=chromosomes, colour=chromosomes)) + geom_point(alpha=0.8, size=2) + ylab("-log10(p-value)") + xlab("Ordered Genotype") + ggtitle("Manhattan Plot") + geom_hline(yintercept=-log10(BF_correction_value), color="red")
```




```{r}
#SNPs <- c(1:N_SNPs)
neg_log_ps <- -log10(p_table_with_chrom_number$pheno_2)
#chromosomes <- factor(p_table_with_chrom_number$chromosome)
df <- data.frame(SNPs,neg_log_ps,p_table_with_chrom_number$chromosome)
ggplot(df, aes(x=SNPs, y=neg_log_ps, group=chromosomes, colour=chromosomes)) + geom_point(alpha=0.8, size=2) + ylab("-log10(p-value)") + xlab("Ordered Genotype") + ggtitle("Manhattan Plot") + geom_hline(yintercept=-log10(BF_correction_value), color="red")

```


```{r}

neg_log_ps <- -log10(p_table_with_chrom_number$pheno_3)
#chromosomes <- factor(p_table_with_chrom_number$chromosome)
df <- data.frame(SNPs,neg_log_ps,p_table_with_chrom_number$chromosome)
ggplot(df, aes(x=SNPs, y=neg_log_ps, group=chromosomes, colour=chromosomes)) + geom_point(alpha=0.8, size=2) + ylab("-log10(p-value)") + xlab("Ordered Genotype") + ggtitle("Manhattan Plot") + geom_hline(yintercept=-log10(BF_correction_value), color="red")

```

```{r}
which(p_value_table[1,] == min(p_value_table[1,]))

```
```{r}
> column_names[16784]
#[1] "rs7726445"
> column_names[16786]
#[1] "rs7731592"
> column_names[16791]
#[1] "rs2161548"
```

```{r}
new_pheno_1_most_sig <- c("rs7726445", "rs7731592", "rs2161548")

new_pheno_1_ids <- getENSG(new_pheno_1_most_sig, mart.snp)[,2];

dat_for_new_pheno_1 = getBM(
  values = c(new_pheno_1_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

new_pheno_1_ids
dat_for_new_pheno_1

```

```{r}
which(p_value_table[2,] == min(p_value_table[2,]))
```

```{r}
column_names[19286]
column_names[19288]
```

```{r}
new_pheno_2_most_sig <- c("rs1129187", "rs10948061")

new_pheno_2_ids <- getENSG(new_pheno_2_most_sig, mart.snp)[,2];

dat_for_new_pheno_2 = getBM(
  values = c(new_pheno_2_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

new_pheno_2_ids
dat_for_new_pheno_2

```

```{r}
which(p_value_table[3,] == min(p_value_table[3,]))
```


```{r}
column_names[41938]
column_names[41942]
column_names[41944]
```

```{r}
new_pheno_3_most_sig <- c("rs11644748", "rs140254902", "rs9652776")

new_pheno_3_ids <- getENSG(new_pheno_3_most_sig, mart.snp)[,2];

dat_for_new_pheno_3 = getBM(
  values = c(new_pheno_3_ids),
  filters = c("ensembl_gene_id"),
  attributes = c("ensembl_gene_id", "external_gene_name", "description"), # what to return
  mart = mart
)

new_pheno_3_ids
dat_for_new_pheno_3

```
