---
title: "Gaudin_QG_Project_Rcode"
author: "Paul Gaudin"
date: "4/30/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


######Import phenotypes and genotypes into tables
```{r}
phenotypes_table <- read.csv("data_files/phenotypes.csv", 
    header = TRUE, row.names = 1)

genotypes_table <- read.csv("data_files/genotypes.csv", 
    header = TRUE, row.names = 1)
````

Check if rownames match 
```{r}
#rownames(phenotypes_table) == rownames(genotypes_table)
```

Get number of SNPs
```{r}
N_SNPs <- ncol(genotypes_table)
```
Get minor alleles in genetype table
```{r}

sample_size <- nrow(genotypes_table)

MAs <- c()  #keep track of minor alleles
MAFs <- c()  #keep track of minor allele frequencies 

for (i in 1:N_SNPs) {
    F_0 <- sum(genotypes_table[, i] == 0)
    F_1 <- sum(genotypes_table[, i] == 1)
    F_2 <- sum(genotypes_table[, i] == 2)
    if (F_0 < F_2) {
        MAFs <- c(MAFs, (F_0 * 2 + F_1)/(sample_size * 2))  #2 for homozygotes, 1 for heterozygotes
        MAs <- c(MAs, F_0)
    } else {
        MAFs <- c(MAFs, (F_2 * 2 + F_1)/(sample_size * 2))  #2 for homozygotes, 1 for heterozygotes
        MAs <- c(MAs, F_2)
    }
}

```

Create Xa and Xd matrices 

```{r}
Xa_matrix <- matrix(, nrow = sample_size, ncol = N_SNPs) 
Xd_matrix <- matrix(, nrow = sample_size, ncol = N_SNPs)

for(col in 1:ncol(genotypes_table)) {
  minor_SNP <- MAs[col]
  for(row in 1:sample_size) { 
    current_allele <- genotypes_table[row, col]
    ##A2A2
    if (current_allele == minor_SNP){ 
      Xa_matrix[row,col] <- 1 
      Xd_matrix[row,col] <- -1
    }
    #A1A2
    else if (current_allele == 1){
      Xa_matrix[row,col] <- 0
      Xd_matrix[row,col] <- 1
    }
    #A1A1
    else{
      Xa_matrix[row,col] <- -1
      Xd_matrix[row,col] <- -1
    } 
  }
}
```

```{r}
write.csv(Xa_matrix, file = "Xa_matrix.csv")
write.csv(Xd_matrix, file = "Xd_matrix.csv")

```



```{r}
pval_calculator <- function(pheno_input, xa_input, xd_input){
    n_samples <- length(xa_input)
    
    X_mx <- cbind(1,xa_input,xd_input)
    
    MLE_beta <- ginv(t(X_mx) %*% X_mx) %*% t(X_mx) %*% pheno_input
    y_hat <- X_mx %*% MLE_beta
  
    SSM <- sum((y_hat - mean(pheno_input))^2)
    SSE <- sum((pheno_input - y_hat)^2)
  
    df_M <- 2
    df_E <- n_samples - 3 
  
    MSM <- SSM / df_M
    MSE <- SSE / df_E
    
    Fstatistic <- MSM / MSE
  
    pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)
    
    return(pval)
}
```

```{r}
library(MASS)
p_value_table <- matrix(, nrow = ncol(phenotypes_table), ncol = N_SNPs)

for(pheno_col in 1:ncol(phenotypes_table)){
  pvals <- rep(0,ncol(Xa_matrix))
  for(i in 1:ncol(Xa_matrix)){
    pvals[i] <- pval_calculator(phenotypes_table[,pheno_col], Xa_matrix[,i], Xd_matrix[,i])
  }
  p_value_table[pheno_col,] <- pvals
}

```

```{r}
write.csv(p_value_table, file = "p_value_table.csv")
```

```{r}

library(ggplot2)

library(ggfortify)

pca_matrix <- read.csv("data_files/genotypes.csv", 
    header = TRUE, row.names = 1)

covars <- read.csv("data_files/covars.csv", 
    header = TRUE, row.names = 1)

```

```{r}
pca_matrix[,ncol(genotypes_table) + 1] <- covars$Population
colnames(pca_matrix)[ncol(genotypes_table) + 1] <- "Population"
autoplot(prcomp(genotypes_table), data = pca_matrix, colour = 'Population')
```

```{r}
pca_matrix[,ncol(genotypes_table) + 2] <- covars$Sex
colnames(pca_matrix)[ncol(genotypes_table) + 2] <- "Sex"
autoplot(prcomp(genotypes_table), data = pca_matrix, colour = "Sex")
```


```{r}
autoplot(prcomp(Xa_matrix), data = pca_matrix, colour = "Population")

```

TSI = -1
CEU = 0
GBR = 1
FIN = 2

```{r}

covar_pheno_vals <- rep(0,length(covars$Population))

for(i in 1:length(covars$Population)){
  if(covars$Population[i] =="TSI"){
    covar_pheno_vals[i] <- -1
  }
  else if(covars$Population[i] =="CEU"){
    covar_pheno_vals[i] <- 0
  }
  else if(covars$Population[i] =="GBR"){
    covar_pheno_vals[i] <- 1
  }
  else{
    covar_pheno_vals[i] <- 2
  }
}

```


```{r}
pval_calculator_covar <- function(xa_input, pheno_input, z_input){ 
  xa_input <- xa_input - 1
  xd_input <- 1 - 2*abs(xa_input)
  n_samples <- length(xa_input)
  Z_mx <- cbind(1,z_input)
  XZ_mx <- cbind(1,xa_input,xd_input,z_input)
  MLE_beta_theta0 <- ginv(t(Z_mx) %*% Z_mx) %*% t(Z_mx) %*% pheno_input 
  MLE_beta_theta1 <- ginv(t(XZ_mx) %*% XZ_mx) %*% t(XZ_mx) %*% pheno_input 
  y_hat_theta0 <- Z_mx %*% MLE_beta_theta0
  y_hat_theta1 <- XZ_mx %*% MLE_beta_theta1
  SSE_theta0 <- sum((pheno_input - y_hat_theta0)^2)
  SSE_theta1 <- sum((pheno_input - y_hat_theta1)^2)
  df_M <- 2
  df_E <- n_samples - 3
  Fstatistic <- ((SSE_theta0-SSE_theta1)/df_M) / (SSE_theta1/df_E)
  pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)
  return(pval)
}
```

```{r}
library(MASS)
p_value_cov_table <- matrix(, nrow = ncol(phenotypes_table), ncol = N_SNPs)

for(pheno_col in 1:ncol(phenotypes_table)){
  #print(pheno_col)
  #pvals <- rep(0,ncol(Xa_matrix))
  for(i in 1:ncol(Xa_matrix)){
    p_value_cov_table[pheno_col,i] <- pval_calculator_covar(Xa_matrix[,i], phenotypes_table[,pheno_col], covar_pheno_vals)
  }
  #p_value_cov_table[pheno_col,] <- pvals
}

```

```{r}
create_qq <- function(pvals){
  N = length(pvals)
  log_sorted_pvals = sort(-log10(pvals))
  N_vector = sort(-log10(seq(0, 1, by = 1/N)))[1:N]
  pval_qqDf <- data.frame(pvals = log_sorted_pvals, N_vector = N_vector)
  qq_plot <- ggplot(pval_qqDf, aes(N_vector, log_sorted_pvals)) + geom_point() + 
    geom_abline(intercept = 0, slope = 1, color = "red")
  return(qq_plot)
}
```

```{r}

qq_1 <- create_qq(p_value_table[1,]) + labs(y = "Computed PVals", x = "")
qq_2 <- create_qq(p_value_table[2,]) + labs(y = "", x = "")
qq_3 <- create_qq(p_value_table[3,]) + labs(y = "", x = "Theor. Pvals")
qq_4 <- create_qq(p_value_table[4,]) + labs(y = "", x = "")
qq_5 <- create_qq(p_value_table[5,]) + labs(y = "", x = "")


```

```{r fig1, fig.height = 1, fig.width=5}
library(ggpubr)
theme_set(theme_pubr())

qq_plots_no_cov <- ggarrange(qq_1, qq_2, qq_3, qq_4, qq_5,
                    labels = c("1", "2", "3", "4", "5"),
                    ncol = 5, nrow = 1)

qq_plots_no_cov

```


```{r}
qq_1_cov <- create_qq(p_value_cov_table[1,]) + labs(y = "Computed PVals", x = "")
qq_2_cov <- create_qq(p_value_cov_table[2,]) + labs(y = "", x = "")
qq_3_cov <- create_qq(p_value_cov_table[3,]) + labs(y = "", x = "Theor. Pvals")
qq_4_cov <- create_qq(p_value_cov_table[4,]) + labs(y = "", x = "")
qq_5_cov <- create_qq(p_value_cov_table[5,]) + labs(y = "", x = "")
```

```{r fig1, fig.height = 1, fig.width=5}

qq_plots_with_cov <- ggarrange(qq_1_cov, qq_2_cov, qq_3_cov, qq_4_cov, qq_5_cov,
                    labels = c("1", "2", "3", "4", "5"),
                    ncol = 5, nrow = 1)

qq_plots_with_cov
```

```{r}


```
